src/hatch/cli/application.py

```diff
@@ -91,10 +92,12 @@ def runner_context(
         *,
         ignore_compat: bool = False,
         display_header: bool = False,
+        keep_env: bool = False,
     ) -> Generator[ExecutionContext, None, None]:
         if self.verbose or len(environments) > 1:
             display_header = True
 
+        self.keep_env = keep_env
```

Comment:
I think we should strive to never modify properties of an instance. Is there another way here?